src/hotspot/share/runtime/mountUnmountDisabler.cpp

```diff
@@ -0,0 +1,450 @@
+ * questions.
+ *
+ */
+
+#include "classfile/javaClasses.inline.hpp"
+#include "prims/jvmtiEventController.hpp"
+#include "prims/jvmtiExport.hpp"
+#include "prims/jvmtiThreadState.inline.hpp"
+#include "runtime/handles.inline.hpp"
+#include "runtime/javaThread.hpp"
+#include "runtime/mountUnmountDisabler.hpp"
+#include "runtime/threadSMR.hpp"
+
+volatile int MountUnmountDisabler::_global_vthread_transition_disable_count = 0;
+volatile int MountUnmountDisabler::_active_disablers = 0;
+bool MountUnmountDisabler::_exclusive_operation_ongoing = false;
+bool MountUnmountDisabler::_notify_jvmti_events = false;
+
+#if INCLUDE_JVMTI
+class JVMTIStartTransition : public StackObj {
+  JavaThread* _current;
+  Handle _vthread;
+  bool _is_mount;
+  bool _is_thread_end;
+ public:
+  JVMTIStartTransition(JavaThread* current, oop vthread, bool is_mount, bool is_thread_end) :
+    _current(current), _vthread(current, vthread), _is_mount(is_mount), _is_thread_end(is_thread_end) {
+    assert(DoJVMTIVirtualThreadTransitions || !JvmtiExport::can_support_virtual_threads(), "sanity check");
+    if (DoJVMTIVirtualThreadTransitions && MountUnmountDisabler::notify_jvmti_events()) {
+      // post VirtualThreadUnmount event before VirtualThreadEnd
+      if (!_is_mount && JvmtiExport::should_post_vthread_unmount()) {
+        JvmtiExport::post_vthread_unmount((jthread)_vthread.raw_value());
+      }
+      if (_is_thread_end && JvmtiExport::should_post_vthread_end()) {
+        JvmtiExport::post_vthread_end((jthread)_vthread.raw_value());
+      }
+    }
+  }
+  ~JVMTIStartTransition() {
+    if (DoJVMTIVirtualThreadTransitions && MountUnmountDisabler::notify_jvmti_events()) {
+      if (_is_thread_end && _current->jvmti_thread_state() != nullptr) {
+        JvmtiExport::cleanup_thread(_current);
+        assert(_current->jvmti_thread_state() == nullptr, "should be null");
+        assert(java_lang_Thread::jvmti_thread_state(_vthread()) == nullptr, "should be null");
+      }
+      if (!_is_mount) {
+        _current->rebind_to_jvmti_thread_state_of(_current->threadObj());
+      }
+    }
+  }
+};
+
+class JVMTIEndTransition : public StackObj {
+  JavaThread* _current;
+  Handle _vthread;
+  bool _is_mount;
+  bool _is_thread_start;
+ public:
+  JVMTIEndTransition(JavaThread* current, oop vthread, bool is_mount, bool is_thread_start) :
+    _current(current), _vthread(current, vthread), _is_mount(is_mount), _is_thread_start(is_thread_start) {
+    assert(DoJVMTIVirtualThreadTransitions || !JvmtiExport::can_support_virtual_threads(), "sanity check");
+    if (DoJVMTIVirtualThreadTransitions && MountUnmountDisabler::notify_jvmti_events()) {
+      if (_is_mount) {
+        _current->rebind_to_jvmti_thread_state_of(_vthread());
+      }
+      DEBUG_ONLY(bool is_virtual = java_lang_VirtualThread::is_instance(_current->jvmti_vthread()));
+      assert((_is_mount && is_virtual) || (!_is_mount && !is_virtual), "wrong identity");
```

Comment:
Nit: Thank you for update! This check can be simplified with:
```
      assert(_is_mount == is_virtual), "wrong identity");
```
But it becomes more obscure though. Feel free to ignore this comment.

Reply:
I applied the suggestion, it reads well to me.